File: 3.x.x/crates-utopia/openapi/src/parser/value.rs
//! Shape checks against a decoded document (PHP `Utopia\OpenAPI\Parser\Value`).

use crate::error::{InvalidSpecification, OpenApiError};
use crate::json::{Json, JsonNumber};
use indexmap::IndexMap;

#[derive(Debug)]
pub struct Value;

impl Value {
    /// PHP `Value::object`. Empty PHP arrays (empty JSON object or array) are objects.
    pub fn object<'a>(
        value: &'a Json,
        location: &str,
    ) -> Result<&'a IndexMap<String, Json>, OpenApiError> {
        match value {
            Json::Object(map) => Ok(map),
            Json::Array(items) if items.is_empty() => Ok(empty_map()),
            _ => Err(InvalidSpecification(format!("Expected an object at {location}")).into()),
        }
    }

    pub fn object_owned(
        value: Json,
        location: &str,
    ) -> Result<IndexMap<String, Json>, OpenApiError> {
        match value {
            Json::Object(map) => Ok(map),
            Json::Array(items) if items.is_empty() => Ok(IndexMap::new()),
            _ => Err(InvalidSpecification(format!("Expected an object at {location}")).into()),
        }
    }

    /// PHP `Value::list`. Empty PHP arrays are lists.
    pub fn list<'a>(value: &'a Json, location: &str) -> Result<&'a [Json], OpenApiError> {
        match value {
            Json::Array(items) => Ok(items),
            Json::Object(map) if map.is_empty() => Ok(&[]),
            _ => Err(InvalidSpecification(format!("Expected a list at {location}")).into()),
        }
    }

    pub fn required_string(
        data: &IndexMap<String, Json>,
        key: &str,
        location: &str,
    ) -> Result<String, OpenApiError> {
        match data.get(key) {
            Some(Json::String(s)) => Ok(s.clone()),
            _ => Err(InvalidSpecification(format!("Expected string {location}/{key}")).into()),
        }
    }

    pub fn optional_string(
        data: &IndexMap<String, Json>,
        key: &str,
    ) -> Result<Option<String>, OpenApiError> {
        match data.get(key) {
            None | Some(Json::Null) => Ok(None),
            Some(Json::String(s)) => Ok(Some(s.clone())),
            _ => Err(InvalidSpecification(format!("Expected '{key}' to be a string")).into()),
        }
    }

    pub fn nullable_int(value: &Json, location: &str) -> Result<Option<i64>, OpenApiError> {
        match value {
            Json::Null => Ok(None),
            Json::Number(n) if n.is_int() => n.as_i64().map(Some).ok_or_else(|| {
                InvalidSpecification(format!("Expected integer at {location}")).into()
            }),
            _ => Err(InvalidSpecification(format!("Expected integer at {location}")).into()),
        }
    }

    pub fn nullable_number(
        value: &Json,
        location: &str,
    ) -> Result<Option<crate::model::JsonNumberOrInt>, OpenApiError> {
        match value {
            Json::Null => Ok(None),
            Json::Number(JsonNumber::Int(i)) => Ok(Some(crate::model::JsonNumberOrInt::Int(*i))),
            Json::Number(JsonNumber::UInt(u)) => match i64::try_from(*u) {
                Ok(i) => Ok(Some(crate::model::JsonNumberOrInt::Int(i))),
                Err(_) => {
                    Err(InvalidSpecification(format!("Expected number at {location}")).into())
                }
            },
            Json::Number(JsonNumber::Float(f)) => {
                Ok(Some(crate::model::JsonNumberOrInt::Float(*f)))
            }
            _ => Err(InvalidSpecification(format!("Expected number at {location}")).into()),
        }
    }

    /// Keys that start with `x-` (case-insensitive), original key preserved.
    pub fn extensions(data: &IndexMap<String, Json>) -> IndexMap<String, Json> {
        data.iter()
            .filter(|(k, _)| k.to_ascii_lowercase().starts_with("x-"))
            .map(|(k, v)| (k.clone(), v.clone()))
            .collect()
    }
}

fn empty_map() -> &'static IndexMap<String, Json> {
    use std::sync::OnceLock;
    static EMPTY: OnceLock<IndexMap<String, Json>> = OnceLock::new();
    EMPTY.get_or_init(IndexMap::new)
}

File: packages/openapi/src/Parser/Value.php
<?php

declare(strict_types=1);

namespace Utopia\OpenAPI\Parser;

use Utopia\OpenAPI\Exception\InvalidSpecification;

/**
 * Shape checks against a decoded document. Every failure names the JSON Pointer
 * location it was reading, so errors can be traced back into the source.
 */
final class Value
{
    /** @return array<string, mixed> */
    public static function object(mixed $value, string $location): array
    {
        if ($value instanceof \stdClass) {
            return (array) $value;
        }
        if (! \is_array($value) || ($value !== [] && array_is_list($value))) {
            throw new InvalidSpecification("Expected an object at {$location}");
        }

        return $value;
    }

    /** @return list<mixed> */
    public static function list(mixed $value, string $location): array
    {
        if (! \is_array($value) || ! array_is_list($value)) {
            throw new InvalidSpecification("Expected a list at {$location}");
        }

        return $value;
    }

    /** @return list<string> */
    public static function stringList(mixed $value, string $location): array
    {
        $values = self::list($value, $location);
        foreach ($values as $index => $item) {
            if (! \is_string($item)) {
                throw new InvalidSpecification("Expected string at {$location}/{$index}");
            }
        }

        return $values;
    }

    /** @param array<string, mixed> $data */
    public static function requiredString(array $data, string $key, string $location): string
    {
        if (! \array_key_exists($key, $data) || ! \is_string($data[$key])) {
            throw new InvalidSpecification("Expected string {$location}/{$key}");
        }

        return $data[$key];
    }

    /** @param array<string, mixed> $data */
    public static function optionalString(array $data, string $key): ?string
    {
        if (! \array_key_exists($key, $data) || $data[$key] === null) {
            return null;
        }
        if (! \is_string($data[$key])) {
            throw new InvalidSpecification("Expected '{$key}' to be a string");
        }

        return $data[$key];
    }

    public static function nullableInt(mixed $value, string $location): ?int
    {
        if ($value === null) {
            return null;
        }
        if (! \is_int($value)) {
            throw new InvalidSpecification("Expected integer at {$location}");
        }

        return $value;
    }

    public static function nullableNumber(mixed $value, string $location): int|float|null
    {
        if ($value === null) {
            return null;
        }
        if (! \is_int($value) && ! \is_float($value)) {
            throw new InvalidSpecification("Expected number at {$location}");
        }

        return $value;
    }

    /**
     * @param  array<int|string, mixed>  $data
     * @return array<string, mixed>
     */
    public static function extensions(array $data): array
    {
        return array_filter(
            $data,
            static fn(string|int $key): bool => \is_string($key) && str_starts_with(strtolower($key), 'x-'),
            ARRAY_FILTER_USE_KEY,
        );
    }
}
